business/suggestionService.js

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Microsoft Corporation and others. Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+const { get, set, sortBy, filter, concat, last, isEmpty } = require('lodash')
+
+class SuggestionService {
+  constructor(definitionService, definitionStore) {
+    this.definitionService = definitionService
+    this.definitionStore = definitionStore
+  }
+
+  /**
+   * Get suggestions for the given coordinates.
+   *
+   * @param {EntityCoordinates} coordinates - The entity for which we are looking for suggestions
+   * @returns {Suggestion} A set of suggested fixes for the definition at the given coordinates.
+   * `null` is returned if no such coordinates are found.
+   */
+  async get(coordinates) {
+    const related = await this._getRelatedDefinitions(coordinates)
+    if (!related) return null
+    const baseSuggestion = this._createBaseSuggestion(coordinates)
+    // create a struct with the strings in it and iterate instead
+    const promises = [
+      await this._collectSuggestionsForField(related, baseSuggestion, 'licensed.declared'),
+      await this._collectSuggestionsForField(related, baseSuggestion, 'described.sourceLocation'),
+      await this._collectSuggestionsForField(related, baseSuggestion, 'described.releaseDate')
+    ]
+    const result = await Promise.all(promises)
+    return last(result.filter(res => !isEmpty(res)))
+  }
+
+  /**
+   * Get related definitions for the given coordinates.
+   * Related here means other definitions for the same component with different revisions
+   *
+   * @param {EntityCoordinates} coordinates - The entity we are looking for related defintions to
+   */
+  async _getRelatedDefinitions(coordinates) {
+    const related = await this.definitionStore.list(coordinates.asRevisionless(), 'definitions')
+    // If the related array only has one entry then return early
+    if (Object.keys(related).length <= 1) {
+      return
+    }
+    const coordinatesList = []
+    related.forEach(element => {
+      coordinatesList.push(element.coordinates)
+    })
+    const validDefinitions = await this.definitionService.getAll(coordinatesList)
```

Comment:
above on L40 we called `list` with `'definitions'` which IIRC should get the actual definition objects. do we still need to collect up the coordinates and get them all?

Reply:
You are right, i've double checked this and seems a duplicate method.

Reply:
After remove it, i've added back this method.
We use getAll() to retrieve whole definition data, because the list() methods returns only coordinates.